src/core/qgsvectorlayerdiagramprovider.cpp

```diff
@@ -78,9 +78,13 @@ QList<QgsLabelFeature *> QgsVectorLayerDiagramProvider::labelFeatures( QgsRender
   QgsFeatureRequest request;
   request.setFilterRect( layerExtent );
   request.setSubsetOfAttributes( attributeNames, mFields );
+  const QgsFeatureFilterProvider *featureFilterProvider = context.featureFilterProvider();
+  if ( featureFilterProvider )
+  {
+    featureFilterProvider->filterFeatures( qobject_cast<QgsVectorLayer *>( mLayer ), request );
```

Comment:
Is it possible that this method is executed in a background thread?

In this case it's dangerous to use the layer directly and the feature source (`mSource`) should be used instead or bad things will happen (the application will crash in flames :boom: ) it looks to me as if this is an error by design in the feature filter provider.

Reply:
It's an API design error, which was unfortunately only flagged after the feature filter work was merged (and we forgot to resolve for 3.0). It is very unsafe, but fortunately it's API which is only us by server for now (and maybe plugins?), and at least for server main thread behaviour is a bit more predictable.

The API needs fixing to remove this use of the layer pointer and replace it with something like the layer uri instead.

Reply:
Anyway, it's not a blocker to this pull request, just something to keep in mind